File: runtime-go/context/layered.go
// Package context / layered.go —— 六层输入的确定性只读投影。见 ch04 §4.4。
//
// LayeredContextEngine 实现 ContextEngine 接口,把 SessionView 投影成 Context。
// **必须是确定性的只读投影**:不写状态、不读时钟、不发起 LLM/Memory 请求。
// 摘要/检索由 Compressor(ch04 §4.5)在上游产出 Event;Assemble 只读 State,
// 并可从 EventStore 展开 WorkingSet 指向的消息原文。
package context

import (
	stdctx "context"
	"fmt"

	"agent-runtime-go/domain"
	"agent-runtime-go/state"
)

// LayeredContextEngine 是 ch04 §4.4 的落地实现。
//
// 依赖:
//   - State:提供 SessionView(WorkingSet / Summaries / MemoryRefs / Progresses)。
//   - Store(可选):当 WorkingSet 里未被 Superseded 的 Turn 需要原文时,从 EventStore 拉。
//   - Instructions:Session 级 system prompt(六层第 1 层)。
//   - Tools:允许的工具集(六层第 6 层)。
//
// Assemble 不调 LLM、不写外部状态;唯一 IO 是确定性地只读 State/EventStore。
type LayeredContextEngine struct {
	State        state.State
	Store        state.EventStore // 用于取 WorkingSet 中未 Superseded 段的原文
	Instructions string           // 相对不变的 system prompt
	Tools        []domain.Tool    // 可用工具集
}

// Assemble 见 ch04 §4.4.1。
//
// 返回的 Messages 顺序:
//  1. system(Instructions)
//  2. system(TaskFrame:Goal + Budget + Constraints)
//  3. system(Progress:当前 Task 的进度快照,包 <task_progress> 标签)
//  4. system(Compressed History:摘要,包 <prior_summary> 标签)
//  5. user/assistant/tool 序列(Working Set 中未 Superseded 的 Turn 原文)
//  6. system(Memory Refs:参考资料,包 <memory_ref> 标签;靠近消息尾部)
func (e *LayeredContextEngine) Assemble(_ stdctx.Context, sessionID, taskID string) (domain.Context, error) {
	view, err := e.State.View(sessionID)
	if err != nil {
		return domain.Context{}, fmt.Errorf("state.View: %w", err)
	}

	msgs := make([]domain.Message, 0, 8)

	// 1. Instructions
	if e.Instructions != "" {
		msgs = append(msgs, domain.Message{Role: "system", Content: e.Instructions})
	}

	// 2. Task Frame
	if task, ok := view.Tasks[taskID]; ok {
		msgs = append(msgs, domain.Message{
			Role:    "system",
			Content: renderTaskFrame(task),
		})
	}

	// 3. Progress
	if progress, ok := view.Progresses[taskID]; ok {
		msgs = append(msgs, domain.Message{
			Role:    "system",
			Content: renderProgress(progress),
		})
	}

	// 4. Compressed History
	//    只放"覆盖 seq 落在 WorkingSet 之外的" Summary,避免与原文重复。
	minSeq := workingSetMinSeq(view.WorkingSet)
	for fromSeq, sum := range view.Summaries {
		if sum.TaskID != "" && taskID != "" && sum.TaskID != taskID {
			continue
		}
		if minSeq > 0 && sum.ToSeq >= minSeq {
			// 该 Summary 覆盖的段还在 WorkingSet 里(且未 Superseded 的部分会作为原文出现)
			// 跳过以避免重复;交给 Working Set 走原文
			_ = fromSeq
			continue
		}
		msgs = append(msgs, domain.Message{
			Role:    "system",
			Content: renderSummary(sum),
		})
	}

	// 5. Working Set:未 Superseded 的 Turn,展开原文。
	//    Superseded 的 Turn 已经在第 3 层作为 Summary 出现。
	if e.Store != nil {
		events, err := e.Store.Load(sessionID)
		if err != nil {
			return domain.Context{}, fmt.Errorf("store.Load: %w", err)
		}
		activeTurns := activeTurnSet(view.WorkingSet, taskID)
		for _, ev := range events {
			if _, ok := activeTurns[ev.TurnID]; !ok {
				continue
			}
			appendTurnMessage(&msgs, ev)
		}
	}

	// 6. Memory Refs:放在 Working Set 之后,使检索证据靠近消息尾部。
	for _, ref := range view.MemoryRefs {
		msgs = append(msgs, domain.Message{
			Role:    "system",
			Content: renderMemoryRef(ref),
		})
	}

	return domain.Context{
		SessionID: sessionID,
		TaskID:    taskID,
		Messages:  msgs,
		Tools:     e.Tools,
	}, nil
}

// ---------- helpers(全是纯函数) ----------

func renderTaskFrame(task domain.Task) string {
	return fmt.Sprintf("<task_frame>\ngoal: %s\nbudget_tokens: %d\n</task_frame>",
		task.Goal, task.Budget.MaxTokens)
}

func renderProgress(p domain.Progress) string {
	out := fmt.Sprintf("<task_progress version=%d updated_at=%q>\ngoal: %s\n",
		p.Version, p.UpdatedAt, p.Goal)
	for _, step := range p.Done {
		out += fmt.Sprintf("done: [%s] %s | %s | %s\n",
			step.Kind, step.Intent, step.Action, step.Observation)
	}
	for _, step := range p.Next {
		out += fmt.Sprintf("next: [%s] %s | %s\n", step.Kind, step.Intent, step.Action)
	}
	for _, loop := range p.Open {
		out += fmt.Sprintf("open: %s (raised_at=%s)\n", loop.Question, loop.RaisedAt)
	}
	out += "</task_progress>"
	return out
}

// renderSummary 把结构化 Summary 序列化成 LLM 可读的文本,包在 <prior_summary> 标签里。
// **保留 at_seq 让 LLM 能引用**,与 §4.6.2 的设计一致。
func renderSummary(s domain.Summary) string {
	out := "<prior_summary>\n"
	if len(s.UserIntents) > 0 {
		out += fmt.Sprintf("user_intents: %v\n", s.UserIntents)
	}
	if len(s.ToolResults) > 0 {
		out += fmt.Sprintf("tool_results: %v\n", s.ToolResults)
	}
	for _, d := range s.DecisionsMade {
		out += fmt.Sprintf("decision(@seq=%d): %s — %s\n", d.AtSeq, d.What, d.Why)
	}
	if len(s.OpenQuestions) > 0 {
		out += fmt.Sprintf("open_questions: %v\n", s.OpenQuestions)
	}
	if len(s.NextActions) > 0 {
		out += fmt.Sprintf("next_actions: %v\n", s.NextActions)
	}
	out += "</prior_summary>"
	return out
}

func renderMemoryRef(r domain.MemoryRef) string {
	return fmt.Sprintf("<memory_ref source=%q score=%.2f>\n%s\n</memory_ref>",
		r.Source, r.Score, r.Content)
}

func workingSetMinSeq(ws []domain.TurnDigest) int64 {
	var min int64
	for _, d := range ws {
		if d.Superseded {
			continue
		}
		if min == 0 || d.FromSeq < min {
			min = d.FromSeq
		}
	}
	return min
}

func activeTurnSet(ws []domain.TurnDigest, taskID string) map[string]struct{} {
	out := map[string]struct{}{}
	for _, d := range ws {
		if d.Superseded {
			continue
		}
		if d.TaskID != "" && taskID != "" && d.TaskID != taskID {
			continue
		}
		out[d.TurnID] = struct{}{}
	}
	return out
}

// appendTurnMessage 只处理"直接对应 role"的 payload,其他事件在 Assemble 层不出现。
func appendTurnMessage(msgs *[]domain.Message, e domain.Event) {
	switch p := e.Payload.(type) {
	case domain.PayloadUserSpoke:
		*msgs = append(*msgs, domain.Message{Role: "user", Content: p.Text})
	case domain.PayloadLLMReplied:
		m := p.Assistant
		if m.Role == "" {
			m.Role = "assistant"
		}
		m.ToolCalls = p.ToolCalls
		*msgs = append(*msgs, m)
	case domain.PayloadToolReturned:
		*msgs = append(*msgs, domain.Message{
			Role: "tool", ToolCallID: p.CallID, Content: p.Content,
		})
	}
}

File: runtime-rs/src/context/layered.rs
//! LayeredContextEngine —— 六层输入的确定性只读投影。见 ch04 §4.4。
//!
//! 与 `runtime-go/context/layered.go` 对齐。
//!
//! **必须是确定性的只读投影**:不写状态、不读时钟、不发起 LLM/Memory 请求。
//! 摘要/检索由 Compressor(ch04 §4.5)在上游产出 Event;Assemble 只读 State,
//! 并可从 EventStore 展开 WorkingSet 指向的消息原文。

use std::collections::HashSet;
use std::sync::{Arc, Mutex};

use super::{ContextEngine, ContextError};
use crate::domain::{
    Context, Event, EventPayload, MemoryRef, Message, Progress, Summary, Task, Tool, TurnDigest,
};
use crate::state::{EventStore, State};

/// ch04 §4.4 的落地实现。
///
/// Assemble 不调 LLM、不写外部状态;唯一 IO 是确定性地只读 State/EventStore。
pub struct LayeredContextEngine {
    pub state: Arc<Mutex<dyn State + Send>>,
    pub store: Option<Arc<Mutex<dyn EventStore + Send>>>,
    pub instructions: String,
    pub tools: Vec<Tool>,
}

impl ContextEngine for LayeredContextEngine {
    fn assemble(&self, session_id: &str, task_id: &str) -> Result<Context, ContextError> {
        let view = {
            let st = self.state.lock().unwrap();
            st.view(session_id).map_err(|e| ContextError(e.0))?
        };

        let mut msgs: Vec<Message> = Vec::new();

        // 1. Instructions
        if !self.instructions.is_empty() {
            msgs.push(Message {
                role: "system".into(),
                content: self.instructions.clone(),
                ..Default::default()
            });
        }

        // 2. Task Frame
        if let Some(task) = view.tasks.get(task_id) {
            msgs.push(Message {
                role: "system".into(),
                content: render_task_frame(task),
                ..Default::default()
            });
        }

        // 3. Progress
        if let Some(progress) = view.progresses.get(task_id) {
            msgs.push(Message {
                role: "system".into(),
                content: render_progress(progress),
                ..Default::default()
            });
        }

        // 4. Compressed History —— 覆盖 seq 落在 WorkingSet 之外的 Summary。
        let min_seq = working_set_min_seq(&view.working_set);
        let mut sum_keys: Vec<i64> = view.summaries.keys().copied().collect();
        sum_keys.sort_unstable();
        for from_seq in sum_keys {
            let sum = &view.summaries[&from_seq];
            if !sum.task_id.is_empty() && !task_id.is_empty() && sum.task_id != task_id {
                continue;
            }
            if min_seq > 0 && sum.to_seq >= min_seq {
                continue;
            }
            msgs.push(Message {
                role: "system".into(),
                content: render_summary(sum),
                ..Default::default()
            });
        }

        // 5. Working Set 展开原文(只读 seq <= view.max_seq 的事件)。
        if let Some(store_arc) = &self.store {
            let events: Vec<Event> = {
                let s = store_arc.lock().unwrap();
                s.load(session_id).map_err(|e| ContextError(e.0))?
            };
            let active = active_turn_set(&view.working_set, task_id);
            for ev in &events {
                if view.max_seq > 0 && ev.seq > view.max_seq {
                    continue;
                }
                if !active.contains(&ev.turn_id) {
                    continue;
                }
                append_turn_message(&mut msgs, ev);
            }
        }

        // 6. Memory Refs:放在 Working Set 之后,使检索证据靠近消息尾部。
        for r in &view.memory_refs {
            msgs.push(Message {
                role: "system".into(),
                content: render_memory_ref(r),
                ..Default::default()
            });
        }

        Ok(Context {
            session_id: session_id.into(),
            task_id: task_id.into(),
            messages: msgs,
            tools: self.tools.clone(),
            ..Default::default()
        })
    }
}

// ---------- helpers (全是纯函数) ----------

fn render_task_frame(task: &Task) -> String {
    format!(
        "<task_frame>\ngoal: {}\nbudget_tokens: {}\n</task_frame>",
        task.goal, task.budget.max_tokens
    )
}

fn render_progress(p: &Progress) -> String {
    let mut out = format!(
        "<task_progress version={} updated_at=\"{}\">\ngoal: {}\n",
        p.version, p.updated_at, p.goal
    );
    for step in &p.done {
        out.push_str(&format!(
            "done: [{:?}] {} | {} | {}\n",
            step.kind, step.intent, step.action, step.observation
        ));
    }
    for step in &p.next {
        out.push_str(&format!(
            "next: [{:?}] {} | {}\n",
            step.kind, step.intent, step.action
        ));
    }
    for open in &p.open {
        out.push_str(&format!(
            "open: {} (raised_at={})\n",
            open.question, open.raised_at
        ));
    }
    out.push_str("</task_progress>");
    out
}

fn render_summary(s: &Summary) -> String {
    let mut out = String::from("<prior_summary>\n");
    if !s.user_intents.is_empty() {
        out.push_str(&format!("user_intents: {:?}\n", s.user_intents));
    }
    if !s.tool_results.is_empty() {
        out.push_str(&format!("tool_results: {:?}\n", s.tool_results));
    }
    for d in &s.decisions_made {
        out.push_str(&format!(
            "decision(@seq={}): {} — {}\n",
            d.at_seq, d.what, d.why
        ));
    }
    if !s.open_questions.is_empty() {
        out.push_str(&format!("open_questions: {:?}\n", s.open_questions));
    }
    if !s.next_actions.is_empty() {
        out.push_str(&format!("next_actions: {:?}\n", s.next_actions));
    }
    out.push_str("</prior_summary>");
    out
}

fn render_memory_ref(r: &MemoryRef) -> String {
    let content = xml_escape(&r.content);
    format!(
        "<memory_ref source=\"{}\" score={:.2}>\n{}\n</memory_ref>",
        xml_escape(&r.source),
        r.score,
        content
    )
}

fn xml_escape(s: &str) -> String {
    s.replace('&', "&amp;")
        .replace('<', "&lt;")
        .replace('>', "&gt;")
}

fn working_set_min_seq(ws: &[TurnDigest]) -> i64 {
    let mut min = 0i64;
    for d in ws {
        if d.superseded {
            continue;
        }
        if min == 0 || d.from_seq < min {
            min = d.from_seq;
        }
    }
    min
}

fn active_turn_set(ws: &[TurnDigest], task_id: &str) -> HashSet<String> {
    let mut out = HashSet::new();
    for d in ws {
        if d.superseded {
            continue;
        }
        if !d.task_id.is_empty() && !task_id.is_empty() && d.task_id != task_id {
            continue;
        }
        out.insert(d.turn_id.clone());
    }
    out
}

fn append_turn_message(msgs: &mut Vec<Message>, e: &Event) {
    match &e.payload {
        EventPayload::UserSpoke(p) => msgs.push(Message {
            role: "user".into(),
            content: p.text.clone(),
            ..Default::default()
        }),
        EventPayload::LLMReplied(p) => {
            let mut m = p.assistant.clone();
            if m.role.is_empty() {
                m.role = "assistant".into();
            }
            m.tool_calls = p.tool_calls.clone();
            msgs.push(m);
        }
        EventPayload::ToolReturned(p) => msgs.push(Message {
            role: "tool".into(),
            tool_call_id: p.call_id.clone(),
            content: p.content.clone(),
            ..Default::default()
        }),
        _ => {}
    }
}
